MauroDataMapper/mdm-explorer
# gh-169 Autocomplete search/select control for query builder

Resolves #169

Custom form control for using autocomplete search and multi-select to pick options for a cohort query. Ideally to be used for scenarios where a large number of option values are present for a field e.g. mapped to a Terminology.

- Combine mat-autocomplete and mat-chip-list components into one
- Implement interfaces required for integration with Angular Forms (ControlValueAccessor) and Angular Material (MatFormFieldControl)
- Integrate into QueryBuilderComponent when data element uses a data type mapped to a Terminology or Code Set
- Render MEQL correctly for autocomplete selected options
- Move QueryBuilderService to DataExplorerModule
- Update QueryBuilderService tests to support new terminology field type
- Add further tests for MeqlPipe
- Add tests to AutocompleteSelectComponent (not exhaustive)
- eslint fixes, including other code to remove warnings
  - QueryBuilderComponent
  - HttpClientService
  - QueryBuilderService
  - MyRequestDetailComponent
  - RequestQueryComponent

# Screenshots

![image](https://user-images.githubusercontent.com/3219480/212121170-222779b1-5fe0-4140-9544-31255a240438.png)

# Testing

To test this feature, you need to make modifications to your mdm-explorer data model via `mdm-ui`:

1. Go to the "modules" data model and view the "Types" tab.
2. Click "Add Data Type", then "Create a New Data Type".
3. Under "Data Type Details", select "Type" as "Terminology" or "Code Set", select a model with many terms, then set any other values as you wish. Click "Submit Data Type".
4. Now select one of the data model classes e.g. "modules > Demographics > Ethnicity (ARIA)", then view the "Elements" tab.
5. Click "Add Data Element", then "Create a New Data Element".
6. Fill in the details and select the data type you created above. Click "Submit Data Element".

Once that is done, go back to `mdm-explorer`:

1. Search for the new data element you created above.
2. Add this data element to a data request (new or existing).
3. Open the request, then edit the cohort/data query.
4. In the query builder, you should now see the option for your data element to be used as a "terminology" type, with autocomplete and select control. Start typing values to search for terms, then add to your query.
